File: src/main/java/com/meridian/lms/entity/Loan.java
package com.meridian.lms.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "loans")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Loan {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "customer_id", nullable = false)
    private User customer;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "loan_type_id", nullable = false)
    private LoanType loanType;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assigned_employee_id")
    private User assignedEmployee;

    @Column(nullable = false, precision = 19, scale = 4)
    private BigDecimal amount;

    @Column(name = "term_months", nullable = false)
    private Integer termMonths;

    @Column(name = "interest_rate", nullable = false, precision = 5, scale = 2)
    private BigDecimal interestRate;

    @Column(name = "monthly_payment", nullable = false, precision = 19, scale = 4)
    private BigDecimal monthlyPayment;

    @Column(name = "total_payable", nullable = false, precision = 19, scale = 4)
    private BigDecimal totalPayable;

    @Column(name = "total_interest", nullable = false, precision = 19, scale = 4)
    private BigDecimal totalInterest;

    @Column(name = "outstanding_balance", nullable = false, precision = 19, scale = 4)
    private BigDecimal outstandingBalance;

    @Enumerated(EnumType.STRING)
    @Builder.Default
    @Column(nullable = false, length = 50)
    private LoanStatus status = LoanStatus.PENDING;

    @Column(columnDefinition = "TEXT")
    private String purpose;

    @Column(name = "rejection_reason", columnDefinition = "TEXT")
    private String rejectionReason;

    @CreationTimestamp
    @Column(name = "applied_at", nullable = false, updatable = false)
    private LocalDateTime appliedAt;

    @Column(name = "reviewed_at")
    private LocalDateTime reviewedAt;

    @Column(name = "approved_at")
    private LocalDateTime approvedAt;

    @Column(name = "completed_at")
    private LocalDateTime completedAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @Version
    @Column(nullable = false)
    private long version;

    public enum LoanStatus {
        PENDING, UNDER_REVIEW, APPROVED, REJECTED, ACTIVE, COMPLETED, DEFAULTED
    }
}

File: frontend/src/types/loan.ts
export type LoanStatus =
  | "PENDING"
  | "UNDER_REVIEW"
  | "APPROVED"
  | "REJECTED"
  | "ACTIVE"
  | "COMPLETED"
  | "DEFAULTED";

export interface LoanType {
  id: number;
  code: string;
  name: string;
  description: string | null;
  minAmount: number;
  maxAmount: number;
  minTermMonths: number;
  maxTermMonths: number;
  baseInterestRate: number;
}

export interface Loan {
  id: number;
  customerId: number;
  customerName: string;
  customerEmail: string;
  loanTypeId: number;
  loanTypeName: string;
  loanTypeCode: string;
  assignedEmployeeId: number | null;
  assignedEmployeeName: string | null;
  amount: number;
  termMonths: number;
  interestRate: number;
  monthlyPayment: number;
  totalPayable: number;
  totalInterest: number;
  outstandingBalance: number;
  status: LoanStatus;
  purpose: string | null;
  rejectionReason: string | null;
  appliedAt: string;
  reviewedAt: string | null;
  approvedAt: string | null;
  completedAt: string | null;
}

export interface LoanApplicationRequest {
  loanTypeCode: string;
  amount: number;
  termMonths: number;
  purpose?: string;
}

export interface ReviewLoanRequest {
  decision: "APPROVE" | "REJECT" | "ASSIGN" | "START_REVIEW";
  reason?: string;
}

export interface LoanFilter {
  status?: LoanStatus;
  loanTypeCode?: string;
  customerId?: number;
  page?: number;
  size?: number;
  sort?: string;
}
